File: algorithms/3_hard/T/00982_triples_with_bitwise_and_equal_to_zero/table.c
__attribute__((no_sanitize("all")))
int countTriplets(const int *restrict nums, const int len) {
    if (len <= 0 || nums == NULL) {
        goto invalid_input;
    }
    int max = INT_MIN;
    #pragma GCC unroll 4
    for (int i = 0; i < len; i += 1) {
        if (nums[i] < 0 || nums[i] >= (1 << 16)) {
            goto invalid_input;
        }
        if (nums[i] > max) {
            max = nums[i];
        }
    }

    size_t table_len = 1u;
    while (table_len <=  (size_t)max) {
        table_len <<= 1u;
    }
    int *restrict table = calloc(table_len, sizeof(*table));
    if (table == NULL) {
        fprintf(stderr, "alloc fail");
        return -1;
    }
    for (int i = 0; i < len; i += 1) {
        #pragma GCC unroll 8
        for (int j = 0; j < len; j += 1) {
            table[(nums[i] & nums[j])] += 1;
        }
    }

    int cnt = 0;
    for (size_t i = 0; i < table_len; i += 1) {
        if (table[i] == 0) {
            continue;
        }
        const int other = (int)i;
        #pragma GCC unroll 4
        for (int j = 0; j < len; j += 1) {
            if ((nums[j] & other) == 0) {
                cnt += table[i];
            }
        }
    }

/* cleanup: */
    free(table);
    return cnt;

invalid_input:
    fprintf(stderr, "invalid input");
    return -1;
}

File: algorithms/3_hard/T/00982_triples_with_bitwise_and_equal_to_zero/table.rs
impl Solution {
    pub fn count_triplets(nums: Vec<i32>) -> i32 {
        debug_assert!(!nums.is_empty() && nums.iter().all(|x| *x >= 0 && *x < 2_i32.pow(16)));
        let max = *nums.iter().max().unwrap();
        let mut table_len = 1_usize;
        while table_len <=  max as usize {
            table_len <<= 1;
        }
        let mut table = vec![0_i32; table_len];
        for &n1 in &nums {
            for &n2 in &nums {
                table[(n1 & n2) as usize] += 1;
            }
        } 

        let mut cnt = 0_i32;
        for (i, &freq) in table.iter().enumerate() {
            if freq == 0 {
                continue;
            }
            let other = i as i32;
            for &num in &nums {
                if (num & other) == 0 {
                    cnt += freq;
                }
            }
        }

        cnt
    }
}
